File: src/GameOfLife.CSharp/Game.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading.Tasks;
using GameOfLife.CSharp.Builder;

namespace GameOfLife.CSharp
{
    public class Game
    {
        private readonly IGameSettings _settings;
        private bool[,] _field;
        private long _generation;

        internal Game(IGameSettings settings)
        {
            _settings = settings ?? throw new ArgumentNullException(nameof(settings));
            _field = new bool[settings.Width, settings.Height];
        }

        public static IGameBuilder CreateBuilder() => new GameOfLifeFluentBuilder();

        public void Prepare()
        {
            _settings.Preset.InitializeField(_field);

            Render(ReadEvents());

            IEnumerable<ChangeEvent> ReadEvents()
            {
                for (var i = 0; i < _field.GetLength(0); i++)
                for (var j = 0; j < _field.GetLength(1); j++)
                {
                    yield return new ChangeEvent(i, j, _field[i, j]);
                }
            }
        }

        public void MakeNextGeneration()
        {
            var events = new ConcurrentBag<ChangeEvent>();
            var newField = _field.Clone() as bool[,];
            Parallel.For(
                0,
                _settings.Width,
                i =>
                {
                    Parallel.For(0, _settings.Height,
                        j =>
                        {
                            bool isAlive = _settings.Ruleset.IsAlive(_field, i, j);
                            if (_field[i, j] != isAlive)
                            {
                                events.Add(new ChangeEvent(i, j, isAlive));
                                newField[i, j] = isAlive;
                            }
                        });
                });

            _field = newField;
            _generation++;

            if (events.Count > 0)
                Render(events);
        }

        private void Render(IEnumerable<ChangeEvent> events)
            => _settings.Render.Render(new GameState(_generation, events));
    }
}

File: src/GameOfLife.FSharp/Game.fs
namespace GameOfLife.FSharp.Game

open System.Collections.Concurrent
open GameOfLife.Abstractions

type ChangeEvent(x, y, isAlive) =
    interface IChangeEvent with
        member this.X = x
        member this.Y = y
        member this.IsAlive = isAlive

type GameState(generation, events) =
    interface IGameState with
        member this.Generation = generation
        member this.Events = events
        
type GameSettings(width, height, preset, render, ruleset) =
    interface IGameSettings with
        member this.Width = width
        member this.Height = height
        member this.Preset = preset
        member this.Render = render
        member this.Ruleset = ruleset

type Game(settings: IGameSettings) =
    let mutable field = Array2D.init settings.Width settings.Height (fun _ _ -> false)
    
    let mutable generation = 0L
    
    let render events =
        let state = GameState(generation, events)
        settings.Render.Render state
    
    let prepare() =
        let w = Array2D.length1 field
        let h = Array2D.length2 field
        
        settings.Preset.InitializeField(field)
        render [| for i in 0..w - 1 do
                  for j in 0..h - 1 do
                    ChangeEvent(i, j, field.[i, j]) |]
        
    let makeNextGen() =
        // single thread
//        let events = ResizeArray<IChangeEvent>()
//        let newField =
//            field
//            |> Array2D.mapi (fun x y value ->
//                let isAlive = settings.Ruleset.IsAlive(field, x , y)
//                match value = isAlive with
//                | true -> value
//                | false -> ChangeEvent(x,y, isAlive) |> events.Add
//                           isAlive)

        // multi thread
        let events = ConcurrentBag<IChangeEvent>()

        let newField = Array2D.copy field
        let width = Array2D.length1 field
        let height = Array2D.length2 field

        let processRow rowIndex rowWidth field =
            for columnIndex in 0..rowWidth - 1 do
                let current = Array2D.get field rowIndex columnIndex
                let isAlive = settings.Ruleset.IsAlive(field, rowIndex, columnIndex)
                if current <> isAlive then
                    newField.[rowIndex, columnIndex] <- isAlive
                    ChangeEvent(rowIndex, columnIndex, isAlive) |> events.Add

        seq { for rowIndex in 0..height - 1 do
                async { processRow rowIndex width field }
        }
        |> Async.Parallel
        |> Async.RunSynchronously
        |> ignore

        field <- newField
        generation <- generation + 1L
        
        events.ToArray() |> render
    
    interface IGame with
        member this.Prepare() = prepare()
        member this.MakeNextGeneration() = makeNextGen()